api/client/cp.go

```diff
@@ -1,57 +1,253 @@
+// so we have to check for a `/` or `.` prefix. Also, in the case of a Windows
+// client, a `:` could be part of an absolute Windows path, in which case it
+// is immediately proceeded by a backslash.
+func splitCpArg(arg string) (container, path string) {
+	if filepath.IsAbs(arg) {
+		// Explicit local absolute path, e.g., `C:\foo` or `/foo`.
+		return "", arg
 	}
-	if statusCode == 404 {
-		return fmt.Errorf("No such container: %v", info[0])
+
+	parts := strings.SplitN(arg, ":", 2)
+
+	switch {
+	case len(parts) == 1:
+		// No `:` in the arg.
+		fallthrough
+	case strings.HasPrefix(parts[0], "."):
+		// Explicit local relative path.
+		return "", arg
+	default:
+		return parts[0], parts[1]
+	}
+}
+
+type containerPathStat struct {
+	Name  string      `json:"name"`
+	Path  string      `json:"path"`
+	Size  int64       `json:"size"`
+	Mode  os.FileMode `json:"mode"`
+	Mtime time.Time   `json:"mtime"`
+}
+
+func (cli *DockerCli) statContainerPath(containerName, path string) (containerPathStat, error) {
+	var stat containerPathStat
+
+	query := make(url.Values, 1)
+	query.Set("path", filepath.ToSlash(path)) // Normalize the paths used in the API.
+
+	urlStr := fmt.Sprintf("/containers/%s/stat-path?%s", containerName, query.Encode())
+
+	response, _, statusCode, err := cli.call("GET", urlStr, nil, nil)
+	if err != nil {
+		return stat, err
+	}
+
+	defer response.Close()
+
+	if statusCode != 200 {
+		return stat, fmt.Errorf("unexpected status code from daemon: %d", statusCode)
+	}
+
+	err = json.NewDecoder(response).Decode(&stat)
+
+	return stat, err
+}
+
+func resolveLocalPath(localPath string) (absPath string, err error) {
+	if absPath, err = filepath.Abs(localPath); err != nil {
+		return
+	}
+
+	return archive.PreserveTrailingDotOrSeparator(absPath, localPath), nil
+}
+
+func (cli *DockerCli) copyFromContainer(srcContainer, srcPath, dstPath string) (err error) {
+	if dstPath != "-" {
+		// Get an absolute destination path.
+		dstPath, err = resolveLocalPath(dstPath)
+		if err != nil {
+			return err
+		}
 	}
+
+	query := make(url.Values, 1)
+	query.Set("path", filepath.ToSlash(srcPath)) // Normalize the paths used in the API.
+
+	urlStr := fmt.Sprintf("/containers/%s/archive-path?%s", srcContainer, query.Encode())
+
+	response, hdr, statusCode, err := cli.call("GET", urlStr, nil, nil)
 	if err != nil {
 		return err
 	}
 
-	hostPath := cmd.Arg(1)
-	if statusCode == 200 {
-		if hostPath == "-" {
-			_, err = io.Copy(cli.out, stream)
-		} else {
-			err = archive.Untar(stream, hostPath, &archive.TarOptions{NoLchown: true})
+	defer response.Close()
+
+	if statusCode != 200 {
+		return fmt.Errorf("unexpected status code from daemon: %d", statusCode)
+	}
+
+	if dstPath == "-" {
+		// Send the response to STDOUT.
+		_, err = io.Copy(os.Stdout, response)
+
+		return err
+	}
+
+	// Prepare source copy info.
+	srcInfo := archive.CopyInfo{
+		Path:   srcPath,
+		Exists: true,
+		IsDir:  hdr.Get("X-Docker-Container-Path-Is-Dir") == "true",
```

Comment:
:( Is it so bad that we can version the API if a change is needed?
I'd rather not hold up both the MUCH imporved cp handling, and being able to cp to container.
The implications of this PR are huge... specific semantics of things that don't exist today seem secondary.